### PR TITLE
files_external: proper user context for sharing

### PR DESCRIPTION
**Pull request by juliusknorr**
This will make sure we also have a proper user context for non-logged in users that access the instance though share links. Otherwise the file system setup doesn't work properly for share links, if either a $home or $user variable is configured for substitution.

Steps to reproduce:
- Configure external storage mounted to / with $user contained in the path
- Try to access file shared via link

### CODE REVIEW

**Review by blizzz — APPROVED** [review]

**Comment by blizzz** [comment]
That's a good catch already. Another use case are internal shares with placeholders in the config. They also don't resolve because of the wrong replacement it seems.

**Comment by scanom** [comment]
Just tested the PR, it solves partially the issues referenced. Shared files with path containing $user variable work fine, but opening then with onlyoffice still don't work (download failed). As before, without $user file opens fine.

**Comment by juliushaertl** [comment]
> That's a good catch already. Another use case are internal shares with placeholders in the config. They also don't resolve because of the wrong replacement it seems.

Hm, then I think it would be better if we pass though the user id that is used to setup the storage (as it seems that it was in stable15 and before). I'll have another look.

**Comment by juliushaertl** [comment]
@blizzz I've pushed another commit so we reuse the user id that is used to setup the mountpoints https://github.com/nextcloud/server/blob/1b85ef4bf2a18f7ee7bca7bcd3d7702362125d2e/lib/private/Files/Filesystem.php#L395

@scanom I don't have a only office test setup, so maybe you can give that another try. At least for collabora this works just fine now.

**Comment by blizzz** [comment]
@juliushaertl let me test it today

**Comment by juliushaertl** [comment]
```
TypeError: Argument 1 passed to OCA\Files_External\Config\UserContext::__construct() must implement interface OCP\IUserSession, instance of Mock_IRequest_d7497604 given, called in /drone/src/apps/files_external/tests/Config/UserPlaceholderHandlerTest.php on line 59
```

Seems the tests also need another look :thinking: Not sure why they don't run yet.

File: apps/files_external/lib/Config/UserContext.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+		$this->session = $session;
+		$this->shareManager = $manager;
+		$this->request = $request;
+	}
+
+	public function getSession(): IUserSession {
+		return $this->session;
+	}
+
+	public function setUser($user): void {
```
Comment (kesselb) [inline apps/files_external/lib/Config/UserContext.php]: Would you mind to check the naming a bit? `$this->user` is returned later as UserId but called User here. I usually try to guess from the name if a method returns a object or an identifier.

**Comment by scanom** [comment]
@juliushaertl PR working flawlessly now, checked both Onlyoffice and Collabora.

**Comment by blizzz** [comment]
Issue persists for me, but it seems the issue is deeper. The wrong storage is used when looking for the file id. Still debugging.

**Comment by blizzz** [comment]
> Issue persists for me, but it seems the issue is deeper. The wrong storage is used when looking for the file id. Still debugging. 

There was an adjustment on the home handler needed, could have seen it way earlier. Stupid me.

**Review by blizzz — APPROVED** [review]

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by MorrisJobke** [comment]
CI fails :/

**Comment by juliushaertl** [comment]
> TypeError: Argument 1 passed to OCA\Files_External\Config\UserContext::__construct() must implement interface OCP\IUserSession, instance of Mock_IRequest_8a9960c5 given, called in /drone/src/apps/files_external/tests/Config/UserPlaceholderHandlerTest.php on line 59

Still not sure why that happens. Could it be that phpunit is not figuring out the proper type hints since it is in the constructor parent class (UserContext)?

**Comment by juliushaertl** [comment]
Tests should be fine now ... stupid typos.

**Review by blizzz — CHANGES_REQUESTED** [review]
my commits got lost

and I think also something of yours, @juliushaertl, because I don't get the share owning user anymore. I don't recall what the change was here and I overwrote my local copy with  this state :(

**Comment by juliushaertl** [comment]
Hm, not sure what happened, because I had your commits locally as I tested them. Github still has them, let try to grab them

**Comment by juliushaertl** [comment]
@blizzz Restored

**Review by blizzz — APPROVED** [review]

**Pull request merged**